### DISCUSSION TITLE
What is the LayeredMetaDataProvider good for?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1107] [Q&A] [answered]
I found that in the workflow java server example in the class [WorkflowServerLauncher](https://github.com/eclipse-glsp/glsp-server/blob/master/examples/org.eclipse.glsp.example.workflow/src/org/eclipse/glsp/example/workflow/launch/WorkflowServerLauncher.java) a LayeredMetaDataProvider is used.

         ElkLayoutEngine.initialize(new LayeredMetaDataProvider());

The class has no comments and also the example is not explained at this point. I never used this component before. What is it good for? Should I use it?

**Comment by tortmayr** [comment, marked as answer]
Hi @rsoika,

sorry for the late reply.
First the TLDR anwser: This class is used in the context of automated server-side layouting. This is a rather specific usecase and I don't think you need this for your OpenBPMN editor.

More context:

GLSP has an optional feature to support automated server-side layouting.  The `ELKEngine` (and `LayeredMetaDataProvider`) is used in this context. 

In general, it is possible to configure a [`LayoutEngine`](https://github.com/eclipse-glsp/glsp-server/blob/a48db7534ae1a729a3141ae50080b9c05ceb7eb5/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/layout/LayoutEngine.java#L21) on the server side which can be use to autolayout the graphical model on the server side before sending it to the client.

We provide a [reusable implementation](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.layout/src/org/eclipse/glsp/layout/ElkLayoutEngine.java) that uses the powerful Eclipse Layout Kernel (https://eclipse.dev/elk/) under the hood.

In the Workflow example we have a layout engine configured that uses the `Layered` ELK algorithm. This is setup with the `LayeredElkMetadataProvider`.

Here is a quick demo:

![layout-demo](https://github.com/eclipse-glsp/glsp/assets/2311075/f66cfd8f-5564-4e83-84d2-7d598fc28683)


Note that the entire server layout/automated layout feature is currently mostly undocumented. We are planning on reiterating/extending the GLSP website documentation once the 2.0. release is finalized and hopefully will also be able to add a section that explains the server-side layouting feature in more detail.

**Reply by rsoika** [reply]
Thanks, yes I understand, and your are right, for the BPMN usecase this is not necessary.
